File: attention/kda_input_proj/torch_ops_extension/custom_ops/csrc/npu_kda_input_proj.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <torch/extension.h>
#include <torch/library.h>

namespace custom {

std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor> construct_kda_input_proj_output_tensors(const at::Tensor &x)
{
    (void)x;
    return std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor>();
}

std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor> npu_kda_input_proj_npu(
    const at::Tensor &x, const at::Tensor &weight_qkv, const at::Tensor &weight_beta, const at::Tensor &weight_gate,
    const at::Tensor &weight_g, const at::Tensor &weight_qkv_scale, bool trans_weight_qkv, bool trans_weight_beta,
    bool trans_weight_gate, bool trans_weight_g)
{
    (void)weight_qkv;
    (void)weight_beta;
    (void)weight_gate;
    (void)weight_g;
    (void)weight_qkv_scale;
    (void)trans_weight_qkv;
    (void)trans_weight_beta;
    (void)trans_weight_gate;
    (void)trans_weight_g;
    return construct_kda_input_proj_output_tensors(x);
}

std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor> npu_kda_input_proj_meta(
    const at::Tensor &x, const at::Tensor &weight_qkv, const at::Tensor &weight_beta, const at::Tensor &weight_gate,
    const at::Tensor &weight_g, const at::Tensor &weight_qkv_scale, bool trans_weight_qkv, bool trans_weight_beta,
    bool trans_weight_gate, bool trans_weight_g)
{
    (void)weight_qkv;
    (void)weight_beta;
    (void)weight_gate;
    (void)weight_g;
    (void)weight_qkv_scale;
    (void)trans_weight_qkv;
    (void)trans_weight_beta;
    (void)trans_weight_gate;
    (void)trans_weight_g;
    return construct_kda_input_proj_output_tensors(x);
}
} // namespace custom

TORCH_LIBRARY_IMPL(custom, PrivateUse1, m)
{
    m.impl("npu_kda_input_proj", &custom::npu_kda_input_proj_npu);
}

TORCH_LIBRARY_IMPL(custom, Meta, m)
{
    m.impl("npu_kda_input_proj", &custom::npu_kda_input_proj_meta);
}

File: attention/kda_input_proj/torch_ops_extension/custom_ops/converter/npu_kda_input_proj.py
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import (
    List,
    Optional,
)

import torch
import torchair
from torchair._ge_concrete_graph.fx2ge_converter import register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec
from torchair.ge import attr


@register_fx_node_ge_converter(torch.ops.custom.npu_kda_input_proj.default)
def convert_npu_kda_input_proj(
    x: Tensor,
    weight_qkv: Tensor,
    weight_beta: Tensor,
    weight_gate: Tensor,
    weight_g: Tensor,
    weight_qkv_scale: Tensor,
    trans_weight_qkv: bool = True,
    trans_weight_beta: bool = True,
    trans_weight_gate: bool = True,
    trans_weight_g: bool = True,
    meta_outputs: List[TensorSpec] = None,
):
    return torchair.ge.custom_op(
        "KdaInputProj",
        inputs={
            "x": x,
            "weight_qkv": weight_qkv,
            "weight_beta": weight_beta,
            "weight_gate": weight_gate,
            "weight_g": weight_g,
            "weight_qkv_scale": weight_qkv_scale,
        },
        attrs={
            "trans_weight_qkv": attr.Bool(trans_weight_qkv),
            "trans_weight_beta": attr.Bool(trans_weight_beta),
            "trans_weight_gate": attr.Bool(trans_weight_gate),
            "trans_weight_g": attr.Bool(trans_weight_g),
        },
        outputs=["qkv", "beta", "gate", "g"],
    )
